src/solvers/smt2_incremental/smt_to_string.cpp

```diff
@@ -41,3 +47,185 @@ std::string smt_to_string(const smt_sortt &sort)
   ss << sort;
   return ss.str();
 }
+
+/// \note The printing algorithm in the `smt_term_to_string_convertert` class is
+///   implemented using an explicit `std::stack` rather than using recursion
+///   and the call stack. This is done in order to ensure we can print smt terms
+///   which are nested arbitrarily deeply without risk of exceeding the maximum
+///   depth of the call stack.
+/// \details
+///   The set of `visit` functions push functions to `output_stack`,
+///   which capture the value to be printed. When called these functions may
+///   either print to the output stream or push further functions to the stack
+///   in the case of nodes in the tree which have child nodes which also need to
+///   be printed.
+///   The `push_output(s)` functions exist as convenience functions to allow
+///   pushing the capturing functions to the stack in the reverse order required
+///   for printing, whilst keeping the visit functions easier to read by keeping
+///   the outputs in reading order and separating the capturing code.
+class smt_term_to_string_convertert : private smt_term_const_downcast_visitort
+{
+private:
+  using output_functiont = std::function<void(std::ostream &)>;
+  std::stack<output_functiont> output_stack;
```

Comment:
I can't help but feel like the concern about call stack depth and the usage of `std::stack` is a case of a bit of premature optimisation... Do we know for sure we could *easily* exhaust the memory through the call stack? Keep in mind that a function call preamble is very lightweight (`push rbp; mov rbp, rsp;  mov DWORD PTR [rbp-4], edi` for an x64 linux system as an example).

If we could exhaust memory that way, what's stopping `std::stack` from exhausting memory? Apologies, but I couldn't find information about the underlying container it uses, but I suspect it could also be vulnerable to the same issue.

Reply:
The size of the call stack has a fixed maximum. It will not grow arbitrarily large. The maximum stack size will also vary depending on which operating system the binary is compiled for. For example the size of the stack on Windows is as little as 1MB. Citation - `https://docs.microsoft.com/en-us/windows/win32/procthread/thread-stack-size` So using recursion to process data structures which may be deeply nested runs the risk of introducing out of stack space bugs which only show on certain platforms.

For a hypothetical example lets say each recursion adds 2 pointers, one for the output stream and one pointing to a term. On a 64 bit architecture those pointers would be 8 bytes each. If this is running with a stack size of 1MB = 2^20, then the maximum recursion depth available would be 2^20 / 8, which works out at 65536. Now this is a large enough number that you aren't likely to reach it with trivial examples but which could be reached with larger examples or if we happen to use an encoding which results in in deeply nested terms. The numbers chosen for this example are fairly arbitrary and are subject to change depending on issues such as -

1. Processor Architecture (for pointer size)
2. Platform (for stack size and calling conventions)
3. Implementation for functions in the algorithm.
4. The optimisations performed by the compiler.

For point 3 the number of arguments to the functions and the number of locally declared variables will affect the rate at which stack memory is used as the recursion gets deeper.

For point 4 it could be possible to write a particular algorithm in such a way that tail call optimisation can be applied and this leads to zero growth in the stack, as the recursion becomes deeper. This could then lead to the situation where the release build does not run out of stack space but the debug build does, due to not applying the optimisation! This could potentially lead to the debug builds being unusable for debugging other unrelated issues.

The unpredictability from above 4 points lead to the potential situation where you could have an example which fails, but does so in surprising ways and only in certain conditions.

The difference with `std::stack`, is that it stores the data contained therein on the heap, which can grow arbitrarily large. Technically the library implementation may have an optimisation to store small container instances on the stack and then expand into the heap as they expand. However the point is that they are only limited by the amount of available memory.

So yes, you could argue that this is a premature optimisation. I am going to argue that this is heading off a potentially difficult to resolve crash bug before it can become a problem.

Reply:
Good stuff, my question was covered adequately - thanks.